[STG-2850] Add Browser Use to v4 migration guide (#2756)
browserbase/stagehand: packages/docs/v4/migrations/browser-use.mdx

### Custom tools

Browser Use's `@tools.action(...)` / `@controller.action(...)` decorators register functions the agent can call. In v4 there's no agent to register them with, so each custom action becomes an ordinary function you call directly in code mode, or one entry in your [tool-calling](#tool-calling) tool set. The function body ports as-is; only the registration goes away.

### Browser configuration

`Browser(...)` (aliased `BrowserSession`) options map onto the browser factory. The common ones:

| Browser Use | Stagehand v4 |
| --- | --- |
| `Browser(headless=False)` | `local_browser.launch(headless=False)` |
| `Browser(cdp_url="http://localhost:9222")` | `local_browser.connect(cdp_url="http://localhost:9222")` |
| `Browser(proxy=ProxySettings(...))` | `proxy` on `local_browser.launch()`, or Browserbase proxies |

> | `Browser(allowed_domains=[...])` / `prohibited_domains=[...]` | `browser.context.set_domain_policy(allowed_domains=..., blocked_domains=...)` |

Author: Good catch — fixed in c7ab9d4. The Python `set_domain_policy` takes a single positional `policy` argument (a `DomainPolicy` TypedDict with `allowed_domains` / `blocked_domains`), so the keyword-arg form would raise `TypeError`. Both table entries now read `set_domain_policy(DomainPolicy(allowed_domains=..., blocked_domains=...))`. Verified against `packages/sdk-python/src/stagehand/browser_context.py` (signature) and `_generated/input_types.py` (the `DomainPolicy` shape); `DomainPolicy` is a public export of the `stagehand` package.

---
title: Migrate Browser Use to v4
sidebarTitle: Migrate Browser Use to v4
icon: 'robot'
---

Browser Use hands a task to an autonomous agent: you write `Agent(task="...", llm=...)`, call `run()`, and the agent decides every click and keystroke on its own. Stagehand v4 has no such object. There is no `Agent`, no `run()`, and nothing that takes a whole task and drives the browser end to end without you.

> So this isn't a rename pass. Browser Use is one primitive; v4 is a browser SDK plus three model-backed steps: [`act()`](/v4/basics/act), [`extract()`](/v4/basics/extract), and [`observe()`](/v4/basics/observe). Moving a flow means deciding, step by step, where a sentence of English is still worth a model call and where a selector does the job for free.

Reviewer: ```suggestion
Browser Use hands a task to an autonomous agent: you write `Agent(task="...", llm=...)`, call `run()`, and the agent picks each action. Stagehand v4 has no equivalent object.

Most of this migration hinges on these principles:
1. There is no `Agent`. Nothing in v4 takes a whole task and drives the browser for you.
2. v4 is a browser SDK plus three model-backed steps. [`act()`](/v4/basics/act), [`extract()`](/v4/basics/extract), and [`observe()`](/v4/basics/observe) take natural-language instructions. Everything else is ordinary browser control.

Porting a flow means writing out the steps the agent used to infer, and deciding for each one whether it needs a model or a selector.
```

<Note>
Browser Use is Python-first, so this guide leads with Python. Stagehand behaves the same way in TypeScript; the [SDK reference](/v4/reference/stagehand) carries each language's naming.
</Note>

## Why there's no Agent

`Agent(task=...).run()` was built for a world where a model couldn't be trusted with the browser on its own, so the framework wrapped it in a loop, showed it the page each step, and asked it to pick one action at a time. Every step was an inference call, whether the page needed judgement or not.

> That trade stopped being worth it. A per-step agent loop is slow, non-deterministic, and expensive precisely where it doesn't need to be: navigating to a URL, clicking a button with a stable selector, reading a table. v4 gives you the discrete tools and leaves the control flow to you.

Reviewer: ```suggestion
Many of those steps do not need a model: navigating to a URL, clicking a button with a stable selector, reading a table. v4 exposes discrete tools and leaves the control flow to you.
```

Two approaches replace the agent:

- **[Code mode](#code-mode)** puts the model in front of the run, not inside it. A coding assistant writes a Stagehand script once; you run that script every time after. Browserbase recommends starting here.
- **[Tool calling](#tool-calling)** keeps a model in the loop at runtime, the way Browser Use does, but drives the browser through the full Stagehand API as its tools instead of one broad task string.

> Either way, `act()`, `extract()`, and `observe()` stay in your toolbox for the steps that genuinely need a model. You just stop handing a model the entire task.

Reviewer: ```suggestion
`act()`, `extract()`, and `observe()` stay available for the steps where a natural-language instruction beats a selector.
```

import asyncio

load_dotenv()

async def main():
    agent = Agent(
        task="Go to news.ycombinator.com and return the title of the top story",
        llm=ChatOpenAI(model="gpt-4.1-mini"),
    )
    history = await agent.run()
    print(history.final_result())

asyncio.run(main())
```

```python Stagehand v4
import asyncio
import os

from stagehand import Stagehand, browserbase

async def main():
    browser = await browserbase.launch(api_key=os.environ["BROWSERBASE_API_KEY"])
    stagehand = await Stagehand.create(browser=browser)

    try:
        page = await browser.context.new_page("https://news.ycombinator.com")
        # A stable selector does this without a model call.
        title = await page.locator(".titleline > a").first().inner_text()
        print(title)
    finally:
        await stagehand.close()
        await browser.close()

asyncio.run(main())
```
</CodeGroup>

The shape of the migration is already visible:

- A browser factory (`browserbase.launch()` or `local_browser.launch()`) replaces the implicit browser inside `Agent`, and `Stagehand.create()` attaches the runtime to it.
- The task string is gone. You write the steps.

> - Where Browser Use would have spent a model call reading the page, a `page.locator()` does it for free. Spend `act()` and `extract()` only where the page needs judgement.

Reviewer: ```suggestion
- Where Browser Use would have spent a model call reading the page, a `page.locator()` reads it directly. Spend `act()` and `extract()` only where the page needs judgement.
```

<Warning>

> Stagehand reads no environment variables of its own. Browser Use auto-loads `.env` and picks up `OPENAI_API_KEY`, `BROWSER_USE_API_KEY`, and friends. In v4 you pass every key explicitly: the Browserbase API key to the factory, and any model key in the `model` option. `load_dotenv()` still works to get values into `os.environ`; nothing reads them for you.

Reviewer: ```suggestion
Stagehand reads no environment variables of its own. Browser Use auto-loads `.env` and picks up `OPENAI_API_KEY`, `BROWSER_USE_API_KEY`, etc. In v4 you pass every key explicitly: the Browserbase API key to the factory, and any model key in the `model` option. `load_dotenv()` still works to get values into `os.environ`; or simply pass saved environment variables using `process.env`.
```

> Stagehand reads no environment variables of its own. Browser Use auto-loads `.env` and picks up `OPENAI_API_KEY`, `BROWSER_USE_API_KEY`, and friends. In v4 you pass every key explicitly: the Browserbase API key to the factory, and any model key in the `model` option. `load_dotenv()` still works to get values into `os.environ`; nothing reads them for you.

Reviewer: ```suggestion
Stagehand reads no environment variables of its own. Browser Use auto-loads `.env` and picks up `OPENAI_API_KEY` and `BROWSER_USE_API_KEY`. In v4 you pass every key explicitly: the Browserbase API key to the factory, and any model key in the `model` option. `load_dotenv()` still populates `os.environ`, but you pass in the values yourself.
```

import { z } from "zod/v4";

const commentSchema = z.object({
  comments: z.array(z.object({ author: z.string(), body: z.string() })),
});

const browser = await browserbase.launch({ apiKey: process.env.BROWSERBASE_API_KEY });
const stagehand = await Stagehand.create({ browser });

try {
  const page = await browser.context.newPage("https://news.ycombinator.com");

  // Deterministic where the page allows it: no inference, no variance.
  await page.locator("a.morelink").first().click();
  await page.waitForLoadState("domcontentloaded");

  // A model call where the page needs judgement.
  await stagehand.act("Open the comments for the story with the most comments");

  const { data } = await stagehand.extract(
    "Extract the top five comments, with each author and body",
    commentSchema,
  );
  console.log(data.comments);
} finally {
  await stagehand.close();
  await browser.close();
}
```
</Tab>
</Tabs>

Generated code should use `page.locator()` and `page.goto()` wherever a selector is stable, and spend a model call only where the page needs judgement. A Browser Use agent couldn't make that split, because every step it ran was an inference call.

## Tool calling

> To keep a model in the loop at runtime, the way Browser Use does, give it the whole Stagehand surface rather than one `task` string. Each method maps to one tool with a narrow contract, so a step names a specific browser operation instead of routing everything through one sentence of English and hoping the agent picks the right action.

Reviewer: ```suggestion
To keep a model in the loop at runtime, give it the whole Stagehand surface instead of one `task` string. Each method becomes one tool with a narrow contract where every step names a specific browser operation.
```

    parameters: z.object({ selector: z.string() }),
    execute: async ({ selector }: { selector: string }) =>
      await page.locator(selector).textContent(),
  },
  act: {
    description: "Perform one action in natural language when no selector is known",
    parameters: z.object({ instruction: z.string() }),
    execute: async ({ instruction }: { instruction: string }) =>
      (await stagehand.act(instruction)).data.message,
  },
  extract: {
    description: "Read structured data off the current page",
    parameters: z.object({ instruction: z.string() }),
    execute: async ({ instruction }: { instruction: string }) =>
      (await stagehand.extract(instruction)).data.extraction,
  },
};
```
</Tab>
</Tabs>

<Tip>
Escalate on `observe()`, never on `act()`. A failed `act()` may already have clicked, submitted, or paid before the error surfaced, so retrying it can repeat the side effect. `observe()` only plans, so retrying it is free. Browser Use's `max_failures` retry loop had the same hazard; keep the retry on the planning step. [Cost optimization](/v4/best-practices/cost-optimization) applies the same idea to model escalation.
</Tip>

## Let a coding assistant do the rest

> Most of the mechanical mapping below is exactly what an assistant is good at. Point it at this page instead of retyping the rules:

Reviewer: ```suggestion
The rest of this guide is a mapping table. Point a coding assistant at this page instead of retyping the rules:
```

```text
Migrate this file from Browser Use to Stagehand v4.

Follow https://docs.stagehand.dev/v4/migrations/browser-use, and use its quick
reference table as the mapping. For each Agent(task=...).run() call, stop and
ask me whether to replace it with a written script (code mode) or a tool-calling
loop, and list every one you find instead of guessing.
```

Set up [AI rules](/v4/first-steps/ai-rules) first, so the assistant stays on the v4 API instead of the Stagehand and Browser Use patterns in its training data.

Then work through the sections below for anything it missed.

## Recommended migration order

1. Get one script constructing and closing cleanly on v4, before porting any behavior. Launch a browser, open a page, close both handles.
2. Replace `Agent(task=...).run()` with [code mode](#code-mode) or [tool calling](#tool-calling). This is the real work, and everything else is mechanical.

> 3. Convert the deterministic steps to `page.locator()` and `page.goto()`, keeping `act()` and `extract()` only where the page needs judgement.

Reviewer: ```suggestion
2. Replace `Agent(task=...).run()` with [code mode](#code-mode) or [tool calling](#tool-calling).
3. Convert the deterministic steps to `page.locator()` and `page.goto()`, keeping `act()` and `extract()` for the steps that need a model.
```

+ stagehand = await Stagehand.create(browser=browser)
+ try:
+     ...  # your steps
+ finally:
+     await stagehand.close()
+     await browser.close()
```
</Tab>

<Tab title="TypeScript">
```diff
+ import { browserbase, Stagehand } from "@browserbasehq/stagehand";
+
+ const browser = await browserbase.launch({ apiKey: process.env.BROWSERBASE_API_KEY });
+ const stagehand = await Stagehand.create({ browser });
+ try {
+   // your steps
+ } finally {
+   await stagehand.close();
+   await browser.close();
+ }
```
</Tab>
</Tabs>

Use `local_browser.launch()` for a browser on your machine, `browserbase.launch({ apiKey })` for a hosted one, and `local_browser.connect({ cdpUrl })` or `browserbase.connect({ apiKey, sessionId })` to attach to one that's already running. `browserbase.launch()` is what enables [server-side caching](/v4/best-practices/caching) and the [Model Gateway](/v4/configuration/models#model-gateway). Stagehand closes only the browsers it launched, so `stagehand.close()` leaves the browser running and you call `browser.close()` yourself. See [browser configuration](/v4/configuration/browser).

### The task string becomes explicit steps

> There's no direct diff here, because a `task` string has no single replacement. That's the whole migration: what the agent inferred each step, you now write out. Read [code mode](#code-mode) and [tool calling](#tool-calling), pick one, and translate the intent of the task into steps.

Reviewer: ```suggestion
There's no direct diff here, because a `task` string has no single replacement. This is the real work, and the rest of this page is mechanical. Read [code mode](#code-mode) and [tool calling](#tool-calling), pick one, and translate the intent of the task into steps.
```

+   variables: { password: process.env.PW },
+ });
+ await stagehand.act("click the login button");
```
</Tab>
</Tabs>

Stagehand exposes only the variable names to the model and substitutes the real values locally. One exception: with [server-side caching](/v4/best-practices/caching) on, variable values travel to the cache service, so turn `cache` off for calls that carry credentials. See [act](/v4/basics/act#secure-your-automations). For Browser Use's TOTP support (`sensitive_data` keys ending in `bu_2fa_code`), generate the code in your own script and pass it as a variable; v4 has no built-in 2FA step.

### Custom tools

Browser Use's `@tools.action(...)` / `@controller.action(...)` decorators register functions the agent can call. In v4 there's no agent to register them with, so each custom action becomes an ordinary function you call directly in code mode, or one entry in your [tool-calling](#tool-calling) tool set. The function body ports as-is; only the registration goes away.

### Browser configuration

`Browser(...)` (aliased `BrowserSession`) options map onto the browser factory. The common ones:

| Browser Use | Stagehand v4 |
| --- | --- |

> | `Browser(headless=False)` | `local_browser.launch({ headless: false })` |

Reviewer: I'd make all of the table into python syntax because that's what the migration guide mentions at the top (and what BU is)

| `tools` / `controller` registry | Your own tool set, see [tool calling](#tool-calling) |
| `history.final_result()` | The return of your last `extract()` |
| `history.structured_output` | `extract().data`, typed by the schema |
| `history.urls()`, `history.errors()` | Your own bookkeeping between steps |
| `use_vision=True` | `screenshot: true` on an `extract()` call |
| `generate_gif=True` | Browserbase [session recording](/v4/configuration/observability) |
| `Browser(...)` / `BrowserSession(...)` | `local_browser.launch()` or `browserbase.launch()` |
| `Browser(cdp_url=...)` | `local_browser.connect({ cdpUrl })` |
| `Browser(allowed_domains=...)` | `browser.context.setDomainPolicy({ allowedDomains })` |
| `Browser(storage_state=...)` | Cookie API, or a Browserbase context |
| `@sandbox(...)` | `browserbase.launch({ apiKey })` |
| `calculate_cost=True`, `history.usage` | `await stagehand.metrics()` |

## Troubleshooting

**`ImportError: cannot import name 'Agent'`.** There is no `Agent` in v4. Replace `Agent(task=...).run()` with [code mode](#code-mode) or [tool calling](#tool-calling).

> **Nothing reads my API key.** Stagehand reads no environment variables. Pass the Browserbase key to `browserbase.launch()` and any model key in the `model` option. `load_dotenv()` only populates `os.environ`; you still pass the values in.

Reviewer: ```suggestion
**`KeyError: 'BROWSERBASE_API_KEY'`.** Stagehand reads no environment variables. Pass the Browserbase key to `browserbase.launch()` and any model key in the `model` option. `load_dotenv()` only populates `os.environ`; you still pass the values in.
```

> **My script has no `history` object to read results from.** There's no run history. The value you'd have read from `history.final_result()` or `history.structured_output` is the return of your last `extract()` call, on `.data`.

Reviewer: ```suggestion
**`AttributeError` on `history` or `final_result()`.** There's no run history. The value you'd have read from `history.final_result()` or `history.structured_output` is the return of your last `extract()` call, on `.data`.
```

> **My custom `@tools.action` function has nowhere to register.** Call it directly in code mode, or add it to your tool set for [tool calling](#tool-calling). Only the decorator goes away; the function body is unchanged.

Reviewer: ```suggestion
**A `@tools.action` function has nowhere to register.** Call it directly in code mode, or add it to your tool set for [tool calling](#tool-calling). Only the decorator goes away; the function body is unchanged.
```

> **A retried step repeats a side effect.** You're retrying `act()`, the same hazard Browser Use's `max_failures` loop had. Retry `observe()` instead and pass the resulting action to `act()` once.

Reviewer: ```suggestion
**A retried step repeats a side effect.** You're retrying `act()`. Retry `observe()` instead and pass the resulting action to `act()` once.
```

> **My generated script uses `Agent` or old Stagehand APIs.** The assistant is drawing on Browser Use and older Stagehand patterns in its training data. Install the rule files from [AI rules](/v4/first-steps/ai-rules).

Reviewer: ```suggestion
**Your generated script uses `Agent` or old Stagehand APIs.** The assistant is drawing on Browser Use and older Stagehand patterns in its training data. Install the rule files from [AI rules](/v4/first-steps/ai-rules).
```
